Specify requested properties for Material Model

anne-glerum opened the pull request:
There are many places where we evaluate the material model to get at one material property. However, many other properties are also computed, like the reaction_terms. This PR specifies in several places that only the viscosity is requested. This should avoid computing the reaction terms and force terms for elasticity and strain dependent rheologies, as both Elasticity and StrainDependent check for `in.requests_property(MaterialProperties::reaction_terms)`.

However, currently request_property() still does
```
      //TODO: Remove this once all callers set requested_properties correctly
      if ((property & MaterialProperties::Property::reaction_terms) != 0)
        return (strain_rate.size() != 0);
```
so if in.strain_rate is filled (like for viscosity) then reaction_terms are still computed. 

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/master/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [ ] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/master/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/master/doc/modules/changes) directory that will inform other users of my change.

gassmoeller commented on source/postprocess/visualization/shear_stress.cc:
```diff
@@ -57,6 +57,9 @@ namespace aspect
         MaterialModel::MaterialModelOutputs<dim> out(n_quadrature_points,
                                                      this->n_compositional_fields());
 
+        // We do not need to compute anything but the viscosity
+        in.requested_properties = MaterialModel::MaterialProperties::viscosity;
```
It would be nice if MaterialModelInputs had a constructor that would take this as input, then you could write above:
```
        MaterialModel::MaterialModelInputs<dim> in(input_data,
                                                   this->introspection(),
                                                   MaterialModel::MaterialProperties::viscosity);
```

Want to give this a try? If not the current form works as well.

tjhei replied:
Yes, that would be a great new variant of reinit() after deprecating the old.

gassmoeller reviewed (commented):
As discussed this works and is a step in the right direction, but does not do anything at the moment because of the compatibility fixes in `MaterialModelInputs<dim>::requests_property`. Still a step forward to setting this correctly.

For the future: It would be nice if MaterialModelInputs had a constructor that would take this as input, then you could write:
```
        MaterialModel::MaterialModelInputs<dim> in(input_data,
                                                   this->introspection(),
                                                   MaterialModel::MaterialProperties::viscosity);
```

Do you want to give this a try? If not the current form works as well.

bangerth commented:
On 7/16/21 10:50 AM, Rene Gassmoeller wrote:
> For the future: It would be nice if MaterialModelInputs had a constructor that 
> would take this as input, then you could write:
> 
> |MaterialModel::MaterialModelInputs<dim> in(input_data, this->introspection(), 
> MaterialModel::MaterialProperties::viscosity); |
> 
> Do you want to give this a try? If not the current form works as well.

Actually, I would prefer if this flag was set on the *output* object 
(MaterialModelOutputs). The *input* object does not need to know what it is 
being used for, but the output object should specify which fields need to be 
filled.

anne-glerum commented:
@bangerth The requested_properties are already a part of MaterialModelInputs, are you saying that should be changed to MaterialModelOutputs? MaterialModels already check the input for the requested properties (although the check doesn't return the right answer for until it's implemented everywhere).

gassmoeller reviewed (approved):
Nice, thanks for the cleanup! I left a suggestion for how to resolve your question.

@bangerth: I see where you are coming from, but as @anne-glerum mentioned this is already established (#3457), this PR only uses the functionality in more places. I see why you say it could go into the outputs object, but I think when I implemented this I had the following two thoughts:
1. Telling the material model what to do is an input operation.
2. Having this in the output object would mean it is handed over as non-const, meaning the material model could modify what properties to compute internally (not likely, but it seems conceptually not ok).

If you want it to be moved I don't feel very strongly about it, but my preference would be an input. A fix would only need a bit of search and replace.

In any case I think we should not delay this PR for it.

bangerth commented:
On 7/20/21 10:02 AM, Rene Gassmoeller wrote:
> @bangerth <https://github.com/bangerth>: I see where you are coming from, but 
> as @anne-glerum <https://github.com/anne-glerum> mentioned this is already 
> established (#3457 <https://github.com/geodynamics/aspect/pull/3457>), this PR 
> only uses the functionality in more places. I see why you say it could go into 
> the outputs object, but I think when I implemented this I had the following 
> two thoughts:
> 
>  1. Telling the material model what to do is an input operation.
>  2. Having this in the output object would mean it is handed over as
>     non-const, meaning the material model could modify what properties to
>     compute internally (not likely, but it seems conceptually not ok).
> 
> If you want it to be moved I don't feel very strongly about it, but my 
> preference would be an input. A fix would only need a bit of search and replace.

No need. I had forgotten that we already had that and not read the patch to 
remind myself of it.

> In any case I think we should not delay this PR for it.

Correct!

anne-glerum commented on source/postprocess/visualization/seismic_anomalies.cc:
```diff
@@ -147,6 +147,11 @@ namespace aspect
                     in.position = fe_values.get_quadrature_points();
 
                     // we do not need the strain rate
+                    // TODO this statement should be changed to
+                    // in.requested_properties = ...
+                    // but I don't know what is actually needed here,
+                    // as it seems only additional output is requested.
+                    // There is no property for that.
                     in.strain_rate.resize(0);
```
I've added the property `additional_outputs` and included it in `all_properties` as well.

gassmoeller commented on source/material_model/interface.cc:
```diff
@@ -370,9 +370,17 @@ namespace aspect
       else
+      {
         this->strain_rate.resize(0);
+        requested_properties = MaterialProperties::equation_of_state_properties |
+                                         MaterialProperties::reaction_terms |
+                                         MaterialProperties::additional_outputs;
```
This way you overwrite the existing value of requested_properties, which is different from the behavior in the if-branch. I think you were looking for only disabling viscosity? If so that would probably work like this:
```suggestion
        requested_properties = requested_properties & ~MaterialProperties::viscosity;
```

anne-glerum replied:
Ah didn't know about `~`, cool!

gassmoeller reviewed (commented):
This looks nearly ready, but it acquired a few conflicts. If you make the one change I mentioned and rebase this should be good to go.

anne-glerum commented:
Fixed the merge conflicts and addressed your comment. 
This PR has become more than I originally described. The next step would be to remove the `compute_strainrate` flag altogether, in favor of the requested_properties, either using the constructor or the current form.

anne-glerum commented:
Hmm, one test still fails. I'll look into it some more.

anne-glerum commented:
I cannot seem to figure out why this last test shear_heating_with_melt is failing. @jdannberg would you maybe have an idea?

bangerth commented:
@anne-glerum Would you mind rebasing this one more time to resolve the merge conflict?

jdannberg commented:
I had a look at the shear_heating_with_melt test. It looks like the changes in the statistics file are very small. Basically, only the last (8th) digit changes, which really could be due to anything. It looks like the biggest changes are in the screen output in the residual of the compositional fields in the last nonlinear iteration of a given time step (so when the residuals are 1e-16, they might be twice than that or half of that).

The test uses the melt simple material model, which is one of the ones you changed, but I honestly do not see how your change could have affected the results (at least the change makes sense to me). In addition, there are other tests that use the same material model and also use operator splitting that have not changed at all. So I would say it's probably fine; at least I do not have any other good ideas.

tjhei commented:
@zjiaqi2018 Do you have an idea what is going on, why the test output changed? Can you test if on current master tests/gmg_mesh_deform_ghost_entries.prm returns for ``RMS, max velocity`` with and without GMG?

tjhei commented:
I talked with @zjiaqi2018 and it seems this prm might be problematic because we have seen free surface + GMG + elasticity to give incorrect solutions (it is in fact currently disabled, but not with a prescribed surface like this test).
That doesn't explain why the output changes here...

tjhei commented:
@anne-glerum I asked @zjiaqi2018 to either find the problem or work around it by disabling/deleting the test temporarily so we can merge this PR.

@zjiaqi2018 It would still be good to know why the changes in this PR change the output. That doesn't make sense to me. Can you maybe confirm that RMS changes with this PR (try rebasing to current master to exclude the possibility that some unrelated change causes the issue).

zjiaqi2018 commented:
I ran gmg_mesh_deform_ghost_entries.prm with the deal.ii master, but got nan in `in.strain_rate`, I don't know what caused the problem.

anne-glerum commented:
I've made some changes in stokes_matrix_free.cc, but they actually shouldn't change anything, because the defaults should be correct. But let's see what the tester says. 

@zjiaqi2018 Where/when did you get in.strain_rate = nan? Do you have a line nr or function name or something?

zjiaqi2018 commented:
> I've made some changes in stokes_matrix_free.cc, but they actually shouldn't change anything, because the defaults should be correct. But let's see what the tester says.
> 
> @zjiaqi2018 Where/when did you get in.strain_rate = nan? Do you have a line nr or function name or something?

![image](https://user-images.githubusercontent.com/36655731/168900091-ff6c81a4-6ba0-467e-922f-6f64128d5dcd.png)

![image](https://user-images.githubusercontent.com/36655731/168900006-b8d6f7b3-e948-4029-b014-3138fcbb6114.png)

anne-glerum commented:
Thank you @zjiaqi2018 ! I suspect that in assembly.cc line 633 `need_viscosity` and line 578-579 `update_strain_rate` are not set correctly for the GMG case. Could you have a look and see whether they should be true in the case of failing test? I will set them to true in the code just to see if the test then succeeds.

anne-glerum commented:
So was this test wrong before this PR? The test fails, but this branch gives the same velocities as the AMG:
```
-   Solving Stokes system... 34+0 iterations.
+   Solving Stokes system... 43+0 iterations.
 
-     RMS, max velocity:         0.142 m/year, 0.448 m/year
+     RMS, max velocity:         0.129 m/year, 0.408 m/year
```

anne-glerum commented:
@tjhei Is the gmg_mesh_deform_ghost_entries.prm test broken? The GMG result isn't the same as for AMG on main (see my message from 20 days ago), but with this PR it is the same (and therefore different from the test results).

tjhei commented:
Could be. @zjiaqi2018 ?

tjhei commented:
Thank you, @anne-glerum this helped! Let me see if I can fix this correctly.

tjhei commented:
@anne-glerum I think you can drop the last commit after you rebase when #4713 is merged.

bangerth commented:
@tjhei suggested merging this before #4725. What is the status of this patch?

tjhei commented:
Needs a rebase and the last commit to be dropped. Then some code reviews as well. At least I focused on the GMG bugs so far.

anne-glerum commented:
I'm done presenting, so I can get back to my PR's. I'll rebase and drop the test commits today or tomorrow. And yes, then it needs a thorough review :)

tjhei reviewed (approved):
Great! Can you please squash your commits?

anne-glerum commented:
@tjhei Anything else I should do?

tjhei reviewed (commented):
This all looks reasonable to me, but I think another pair of eyes would be good.

gassmoeller commented on include/aspect/material_model/interface.h:
```diff
@@ -203,6 +203,9 @@ namespace aspect
         entropy_derivative_temperature = 256,
         reaction_terms                 = 512,
+        reaction_rates                 = 1024,
+        force_terms                    = 2048,
```
I am not so sure about making this one separate from `additional_outputs`. Structurally it is a subset of `additional_outputs`. I understand we may want to ask for it separately, but in the current form there is a problem: Someone could ask for all `additional_outputs` and then assume all additional outputs will be computed, while `force_terms` will not be computed if not requested. Is there a reason we need this as a separate entry?

gassmoeller commented on source/material_model/interface.cc:
```diff
@@ -283,7 +283,12 @@ namespace aspect
+          requested_properties = MaterialProperties::Property(requested_properties & ~MaterialProperties::viscosity);
+        }
+      else
+        requested_properties = requested_properties | MaterialProperties::viscosity;
```
Yes, I think this line is not needed.

gassmoeller reviewed (approved):
I would say let's get this merged and handle the few small comments in a follow-up. Because this PR touches so many files the risk of adding new conflicts soon is pretty large.

gassmoeller commented:
And thanks for all the work @anne-glerum!
